src/robot/libraries/Telnet.py

```diff
@@ -491,8 +531,17 @@ def __init__(self, host=None, port=23, timeout=3.0, newline='CRLF',
                  encoding='UTF-8', encoding_errors='ignore',
                  default_log_level='INFO', window_size=None, environ_user=None,
                  terminal_emulation=False, terminal_type=None,
-                 telnetlib_log_level='TRACE'):
-        telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+                 telnetlib_log_level='TRACE', connection_timeout=None):
+        if connection_timeout is None:
+            telnetlib.Telnet.__init__(self, host, int(port) if port else 23)
+        else:
+            try:
+                telnetlib.Telnet.__init__(self, host,
+                                          int(port) if port else 23,
+                                          connection_timeout)
+            except socket.timeout:
+                raise AssertionError("Connection timeout after %s."
+                                     % secs_to_timestr(connection_timeout))
```

Comment:
We managed to get socket.timeout by binding a server to localhost port, but not starting to listen:
python -c "import socket; ss = socket.socket(socket.AF_INET, socket.SOCK_STREAM); ss.bind(('127.0.0.1',2424)); raw_input()"

Then opening a connection to 127.0.0.1, 2424 results in socket.timeout on our machines.